helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -556,15 +644,55 @@ public String create(final String path, Object datat, final List<ACL> acl, final
     }
     long startT = System.currentTimeMillis();
     try {
-      final byte[] data = datat == null ? null : serialize(datat, path);
-      checkDataSizeLimit(data);
-      String actualPath = retryUntilConnected(new Callable<String>() {
-        @Override
-        public String call() throws Exception {
-          return getConnection().create(path, data, acl, mode);
+      final byte[] dataBytes = dataObject == null ? null : serialize(dataObject, path);
+      checkDataSizeLimit(dataBytes);
+
+      /*
```

Comment:
I feel it would be much cleaner if you split the session check and mode check related logic to a separate method. All the comments could be much simpler as well.